src/ORM/Marshaller.php

```diff
@@ -282,22 +282,27 @@ protected function _belongsToMany(Association $assoc, array $data, $options = []
 
         $primaryKey = array_flip($assoc->target()->schema()->primaryKey());
         $records = [];
+        $conditions = [];
+        $primaryCount = count($primaryKey);
 
         foreach ($data as $i => $row) {
             if (array_intersect_key($primaryKey, $row) === $primaryKey) {
-                if (!isset($query)) {
-                    $primaryCount = count($primaryKey);
-                    $query = $assoc->find();
-                }
                 $keys = array_intersect_key($row, $primaryKey);
                 if (count($keys) === $primaryCount) {
-                    $query->orWhere($keys);
+                    $conditions[] = $keys;
                 }
             } else {
                 $records[$i] = $this->one($row, $options);
             }
         }
 
+        if (!empty($conditions)) {
+            $query = $assoc->target()->find();
```

Comment:
Do we not want association filtering conditions applied? Seems like that could result in incorrect queries sometimes as well.